File: examples/csharp/BloxApiClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace BloxApiExample
{
    /// <summary>
    /// Client for interacting with the BloxAPI
    /// </summary>
    public class BloxApiClient
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;
        private readonly JsonSerializerOptions _jsonOptions;

        /// <summary>
        /// Create a new BloxAPI client
        /// </summary>
        /// <param name="baseUrl">Base URL for the BloxAPI instance</param>
        public BloxApiClient(string baseUrl = "http://localhost:5000")
        {
            _baseUrl = baseUrl;
            _httpClient = new HttpClient();
            _httpClient.DefaultRequestHeaders.Accept.Add(
                new MediaTypeWithQualityHeaderValue("application/json"));
                
            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = true
            };
        }

        /// <summary>
        /// Get information about a Roblox user
        /// </summary>
        /// <param name="userId">The Roblox user ID</param>
        /// <returns>User information</returns>
        public async Task<JsonElement> GetUserAsync(int userId)
        {
            return await GetAsync($"/api/users/{userId}");
        }

        /// <summary>
        /// Search for Roblox users by keyword
        /// </summary>
        /// <param name="keyword">The search keyword</param>
        /// <param name="limit">Maximum number of results (default: 10)</param>
        /// <returns>Search results</returns>
        public async Task<JsonElement> SearchUsersAsync(string keyword, int limit = 10)
        {
            var queryParams = new Dictionary<string, string>
            {
                { "keyword", keyword },
                { "limit", limit.ToString() }
            };
            
            return await GetAsync("/api/users/search", queryParams);
        }

        /// <summary>
        /// Get information about a Roblox game by universe ID
        /// </summary>
        /// <param name="universeId">The Roblox universe ID</param>
        /// <returns>Game information</returns>
        public async Task<JsonElement> GetGameAsync(int universeId)
        {
            return await GetAsync($"/api/games/{universeId}");
        }

        /// <summary>
        /// Get analytics for a Roblox game
        /// </summary>
        /// <param name="universeId">The Roblox universe ID</param>
        /// <param name="metricType">The metric type (e.g., "visits", "revenue")</param>
        /// <param name="timeFrame">Time frame for analytics (default: "past7days")</param>
        /// <returns>Game analytics</returns>
        public async Task<JsonElement> GetGameAnalyticsAsync(int universeId, string metricType, string timeFrame = "past7days")
        {
            var queryParams = new Dictionary<string, string>
            {
                { "time_frame", timeFrame }
            };
            
            return await GetAsync($"/api/analytics/games/{universeId}/{metricType}", queryParams);
        }

        /// <summary>
        /// Search the Roblox catalog
        /// </summary>
        /// <param name="keyword">The search keyword</param>
        /// <param name="category">Optional category filter</param>
        /// <param name="limit">Maximum number of results (default: 25)</param>
        /// <returns>Search results</returns>
        public async Task<JsonElement> SearchCatalogAsync(string keyword, string category = null, int limit = 25)
        {
            var queryParams = new Dictionary<string, string>
            {
                { "keyword", keyword },
                { "limit", limit.ToString() }
            };
            
            if (!string.IsNullOrEmpty(category))
            {
                queryParams.Add("category", category);
            }
            
            return await GetAsync("/api/catalog/search", queryParams);
        }

        /// <summary>
        /// Get information about a Roblox asset
        /// </summary>
        /// <param name="assetId">The Roblox asset ID</param>
        /// <returns>Asset information</returns>
        public async Task<JsonElement> GetAssetAsync(int assetId)
        {
            return await GetAsync($"/api/assets/{assetId}");
        }

        /// <summary>
        /// Get detailed information about a Roblox asset
        /// </summary>
        /// <param name="assetId">The Roblox asset ID</param>
        /// <param name="includeBundles">Whether to include bundles containing this asset</param>
        /// <returns>Detailed asset information</returns>
        public async Task<JsonElement> GetAssetInfoAsync(int assetId, bool includeBundles = false)
        {
            var queryParams = new Dictionary<string, string>();
            
            if (includeBundles)
            {
                queryParams.Add("include_bundles", "true");
            }
            
            return await GetAsync($"/api/assets/{assetId}/info", queryParams);
        }

        /// <summary>
        /// Get information about a Roblox badge
        /// </summary>
        /// <param name="badgeId">The Roblox badge ID</param>
        /// <returns>Badge information</returns>
        public async Task<JsonElement> GetBadgeAsync(int badgeId)
        {
            return await GetAsync($"/api/badges/{badgeId}");
        }

        /// <summary>
        /// Check moderation status of Roblox content
        /// </summary>
        /// <param name="contentType">The content type (e.g., "asset", "game")</param>
        /// <param name="contentId">The content ID</param>
        /// <returns>Moderation status</returns>
        public async Task<JsonElement> CheckContentModerationAsync(string contentType, int contentId)
        {
            return await GetAsync($"/api/moderation/content/{contentType}/{contentId}");
        }

        /// <summary>
        /// Make a GET request to the API
        /// </summary>
        /// <param name="endpoint">API endpoint</param>
        /// <param name="queryParams">Optional query parameters</param>
        /// <returns>Response as JsonElement</returns>
        private async Task<JsonElement> GetAsync(string endpoint, Dictionary<string, string> queryParams = null)
        {
            string url = $"{_baseUrl}{endpoint}";
            
            if (queryParams != null && queryParams.Count > 0)
            {
                var queryString = new StringBuilder("?");
                foreach (var param in queryParams)
                {
                    if (queryString.Length > 1)
                    {
                        queryString.Append("&");
                    }
                    queryString.Append($"{Uri.EscapeDataString(param.Key)}={Uri.EscapeDataString(param.Value)}");
                }
                url += queryString.ToString();
            }
            
            HttpResponseMessage response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();
            
            string responseBody = await response.Content.ReadAsStringAsync();
            return JsonDocument.Parse(responseBody).RootElement;
        }
    }
    
    /// <summary>
    /// Example program to demonstrate using the BloxApiClient
    /// </summary>
    public class Program
    {
        public static async Task Main(string[] args)
        {
            // Create a client
            var client = new BloxApiClient("http://localhost:5000");
            
            try
            {
                // Get user information for Roblox user ID 1
                Console.WriteLine("Getting user information for Roblox founder...");
                var userInfo = await client.GetUserAsync(1);
                Console.WriteLine(JsonSerializer.Serialize(userInfo, new JsonSerializerOptions { WriteIndented = true }));
                Console.WriteLine();
                
                // Search for users with keyword "Roblox"
                Console.WriteLine("Searching for users with keyword \"Roblox\"...");
                var searchResults = await client.SearchUsersAsync("Roblox", 5);
                Console.WriteLine(JsonSerializer.Serialize(searchResults, new JsonSerializerOptions { WriteIndented = true }));
                Console.WriteLine();
                
                // Get game information for Natural Disaster Survival (189707)
                Console.WriteLine("Getting game information for Natural Disaster Survival...");
                var gameInfo = await client.GetGameAsync(189707);
                Console.WriteLine(JsonSerializer.Serialize(gameInfo, new JsonSerializerOptions { WriteIndented = true }));
                Console.WriteLine();
                
                // Get visits analytics for the game
                Console.WriteLine("Getting visits analytics for the game...");
                var analytics = await client.GetGameAnalyticsAsync(189707, "visits");
                Console.WriteLine(JsonSerializer.Serialize(analytics, new JsonSerializerOptions { WriteIndented = true }));
                Console.WriteLine();
                
                // Search for dominus hats in the catalog
                Console.WriteLine("Searching for dominus hats in the catalog...");
                var catalogResults = await client.SearchCatalogAsync("dominus", "hats");
                Console.WriteLine(JsonSerializer.Serialize(catalogResults, new JsonSerializerOptions { WriteIndented = true }));
                Console.WriteLine();
                
                // Check moderation status of an asset
                Console.WriteLine("Checking moderation status of an asset...");
                var moderationStatus = await client.CheckContentModerationAsync("asset", 1818);
                Console.WriteLine(JsonSerializer.Serialize(moderationStatus, new JsonSerializerOptions { WriteIndented = true }));
            }
            catch (HttpRequestException e)
            {
                Console.WriteLine($"Error making API request: {e.Message}");
            }
            catch (JsonException e)
            {
                Console.WriteLine($"Error parsing JSON response: {e.Message}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Unexpected error: {e.Message}");
            }
        }
    }
}

File: examples/java/BloxApiClient.java
package com.example.bloxapi;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Client for interacting with the BloxAPI
 */
public class BloxApiClient {
    private final HttpClient httpClient;
    private final String baseUrl;
    private final ObjectMapper objectMapper;
    
    /**
     * Create a new BloxAPI client
     * 
     * @param baseUrl Base URL for the BloxAPI instance
     */
    public BloxApiClient(String baseUrl) {
        this.baseUrl = baseUrl;
        this.httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        this.objectMapper = new ObjectMapper();
    }
    
    /**
     * Create a new BloxAPI client with default base URL
     */
    public BloxApiClient() {
        this("http://localhost:5000");
    }
    
    /**
     * Get information about a Roblox user
     * 
     * @param userId The Roblox user ID
     * @return User information as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getUser(int userId) throws IOException, InterruptedException {
        return get("/api/users/" + userId);
    }
    
    /**
     * Search for Roblox users by keyword
     * 
     * @param keyword The search keyword
     * @param limit Maximum number of results (default: 10)
     * @return Search results as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode searchUsers(String keyword, int limit) throws IOException, InterruptedException {
        Map<String, String> params = new HashMap<>();
        params.put("keyword", keyword);
        params.put("limit", String.valueOf(limit));
        
        return get("/api/users/search", params);
    }
    
    /**
     * Get information about a Roblox game by universe ID
     * 
     * @param universeId The Roblox universe ID
     * @return Game information as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getGame(int universeId) throws IOException, InterruptedException {
        return get("/api/games/" + universeId);
    }
    
    /**
     * Get analytics for a Roblox game
     * 
     * @param universeId The Roblox universe ID
     * @param metricType The metric type (e.g., "visits", "revenue")
     * @param timeFrame Time frame for analytics (default: "past7days")
     * @return Game analytics as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getGameAnalytics(int universeId, String metricType, String timeFrame) 
            throws IOException, InterruptedException {
        Map<String, String> params = new HashMap<>();
        params.put("time_frame", timeFrame);
        
        return get("/api/analytics/games/" + universeId + "/" + metricType, params);
    }
    
    /**
     * Search the Roblox catalog
     * 
     * @param keyword The search keyword
     * @param category Optional category filter
     * @param limit Maximum number of results (default: 25)
     * @return Search results as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode searchCatalog(String keyword, String category, int limit) 
            throws IOException, InterruptedException {
        Map<String, String> params = new HashMap<>();
        params.put("keyword", keyword);
        params.put("limit", String.valueOf(limit));
        
        if (category != null && !category.isEmpty()) {
            params.put("category", category);
        }
        
        return get("/api/catalog/search", params);
    }
    
    /**
     * Get information about a Roblox asset
     * 
     * @param assetId The Roblox asset ID
     * @return Asset information as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getAsset(int assetId) throws IOException, InterruptedException {
        return get("/api/assets/" + assetId);
    }
    
    /**
     * Get detailed information about a Roblox asset
     * 
     * @param assetId The Roblox asset ID
     * @param includeBundles Whether to include bundles containing this asset
     * @return Detailed asset information as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getAssetInfo(int assetId, boolean includeBundles) 
            throws IOException, InterruptedException {
        Map<String, String> params = new HashMap<>();
        
        if (includeBundles) {
            params.put("include_bundles", "true");
        }
        
        return get("/api/assets/" + assetId + "/info", params);
    }
    
    /**
     * Get information about a Roblox badge
     * 
     * @param badgeId The Roblox badge ID
     * @return Badge information as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode getBadge(int badgeId) throws IOException, InterruptedException {
        return get("/api/badges/" + badgeId);
    }
    
    /**
     * Check moderation status of Roblox content
     * 
     * @param contentType The content type (e.g., "asset", "game")
     * @param contentId The content ID
     * @return Moderation status as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    public JsonNode checkContentModeration(String contentType, int contentId) 
            throws IOException, InterruptedException {
        return get("/api/moderation/content/" + contentType + "/" + contentId);
    }
    
    /**
     * Make a GET request to the API
     * 
     * @param endpoint API endpoint
     * @return Response as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    private JsonNode get(String endpoint) throws IOException, InterruptedException {
        return get(endpoint, null);
    }
    
    /**
     * Make a GET request to the API with query parameters
     * 
     * @param endpoint API endpoint
     * @param params Query parameters
     * @return Response as JsonNode
     * @throws IOException If an I/O error occurs
     * @throws InterruptedException If the operation is interrupted
     */
    private JsonNode get(String endpoint, Map<String, String> params) 
            throws IOException, InterruptedException {
        String url = baseUrl + endpoint;
        
        if (params != null && !params.isEmpty()) {
            String queryString = params.entrySet().stream()
                .map(entry -> URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8) + "=" + 
                              URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8))
                .collect(Collectors.joining("&"));
            
            url += "?" + queryString;
        }
        
        HttpRequest request = HttpRequest.newBuilder()
                .GET()
                .uri(URI.create(url))
                .header("Accept", "application/json")
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() >= 200 && response.statusCode() < 300) {
            return objectMapper.readTree(response.body());
        } else {
            throw new IOException("HTTP error: " + response.statusCode() + " - " + response.body());
        }
    }
    
    /**
     * Example program to demonstrate using the BloxApiClient
     */
    public static void main(String[] args) {
        try {
            BloxApiClient client = new BloxApiClient();
            
            // Get user information for Roblox user ID 1
            System.out.println("Getting user information for Roblox founder...");
            JsonNode userInfo = client.getUser(1);
            System.out.println(userInfo.toPrettyString());
            System.out.println();
            
            // Search for users with keyword "Roblox"
            System.out.println("Searching for users with keyword \"Roblox\"...");
            JsonNode searchResults = client.searchUsers("Roblox", 5);
            System.out.println(searchResults.toPrettyString());
            System.out.println();
            
            // Get game information for Natural Disaster Survival (189707)
            System.out.println("Getting game information for Natural Disaster Survival...");
            JsonNode gameInfo = client.getGame(189707);
            System.out.println(gameInfo.toPrettyString());
            System.out.println();
            
            // Get visits analytics for the game
            System.out.println("Getting visits analytics for the game...");
            JsonNode analytics = client.getGameAnalytics(189707, "visits", "past7days");
            System.out.println(analytics.toPrettyString());
            System.out.println();
            
            // Search for dominus hats in the catalog
            System.out.println("Searching for dominus hats in the catalog...");
            JsonNode catalogResults = client.searchCatalog("dominus", "hats", 25);
            System.out.println(catalogResults.toPrettyString());
            System.out.println();
            
            // Check moderation status of an asset
            System.out.println("Checking moderation status of an asset...");
            JsonNode moderationStatus = client.checkContentModeration("asset", 1818);
            System.out.println(moderationStatus.toPrettyString());
            
        } catch (IOException | InterruptedException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}